tests/DidCache.spec.ts

```diff
@@ -0,0 +1,126 @@
+    return (this.bufs.length - 1).toString();
+  }
+
+  public async read (address: string): Promise<Buffer> {
+    const idx = +address;
+    return this.bufs[idx];
+  }
+}
+
+function createCreateOpBuf(): Buffer {
+  const createPayload = Base58.encode(Buffer.from(JSON.stringify(didDocJson)));
+  const createOpRequest = {
+    createPayload,
+    signature: 'signature',
+    proofOfWork: 'proof of work'
+  };
+  return Buffer.from(JSON.stringify(createOpRequest));
+}
+
+async function createRootedOp(opBuf: Buffer, cas: Cas): Promise<WriteOperation> {
+  const batch: Buffer[] = [ opBuf ];
+  const batchBuffer = Buffer.from(JSON.stringify(batch));
+  const batchFileAddress = await cas.write(batchBuffer);
+  const op = WriteOperation.create(opBuf, 0, 0, batchFileAddress);
+  return op;
+}
+
+describe('DidCache', () => {
+
+  it('should return non-null url for create op', async () => {
+    const dummyCas = new DummyCas();
+    const didCache = new DidCache(dummyCas, didDocumentUpdate, didDocumentCreate);
+    const createOp = await createRootedOp(createCreateOpBuf(), dummyCas);
+    expect(didCache.apply(createOp)).not.toBeNull();
+  });
+
+  it('first(did) should be did', async () => {
+    const dummyCas = new DummyCas();
+    const didCache = new DidCache(dummyCas, didDocumentUpdate, didDocumentCreate);
+    const createOp = await createRootedOp(createCreateOpBuf(), dummyCas);
+    const createRet = didCache.apply(createOp);
+    expect(createRet).not.toBeNull();
+
+    const did = createRet as string;
+    const firstVersion = await didCache.first(did);
+    expect(firstVersion).toBe(did);
+  });
+
+  it('last(did) should be did', async() => {
+    const dummyCas = new DummyCas();
+    const didCache = new DidCache(dummyCas, didDocumentUpdate, didDocumentCreate);
+    const createOp = await createRootedOp(createCreateOpBuf(), dummyCas);
+    const createRet = didCache.apply(createOp);
+    expect(createRet).not.toBeNull();
+
+    const did = createRet as string;
+    expect(didCache.last(did)).toBe(did);
+  });
+
+  it('prev(did) should be null', async() => {
+    const dummyCas = new DummyCas();
+    const didCache = new DidCache(dummyCas, didDocumentUpdate, didDocumentCreate);
+    const createOp = await createRootedOp(createCreateOpBuf(), dummyCas);
+    const createRet = didCache.apply(createOp);
+    expect(createRet).not.toBeNull();
+
+    const did = createRet as string;
+    const prev = await didCache.prev(did);
+    expect(prev).toBeNull();
+  });
+
+  it('should resolve created did', async () => {
+    const dummyCas = new DummyCas();
+    const didCache = new DidCache(dummyCas, didDocumentUpdate, didDocumentCreate);
+    const createOp = await createRootedOp(createCreateOpBuf(), dummyCas);
+    const did = didCache.apply(createOp);
+    expect(did).not.toBeNull();
+
+    const resolvedDid = await didCache.resolve(did as string);
+    // TODO: can we get the raw json from did? if so, we can write a better test.
```

Comment:
Isn't DidDocument already an object? Or are you asking for just the JSON string? Feel free to add to the library, but I don't see how it will benefit the test.